tests/test_algorithm.py

```diff
@@ -2018,3 +2025,256 @@ def test_remove_data(self):
+    """
+    @classmethod
+    def setUpClass(cls):
+        start_date = pd.Timestamp('2015-01-05', tz='UTC')
+        cls.first_end_date = pd.Timestamp('2015-01-07', tz='UTC')
+
+        # Start date index
+        sdi = trading_days.get_loc(start_date)
+
+        cls.metadata = make_jagged_equity_info(
+            num_assets=3,
+            start_date=start_date,
+            first_end=cls.first_end_date,
+            frequency=trading_day,
+            periods_between_ends=2,
+        )
+
+        # Add an auto close date column, where liquidation occurs the day after
+        # the end date.
+        cls.metadata['auto_close_date'] = \
+            cls.metadata['end_date'] + trading_day
+
+        # All equity prices start at 1 and increment each day.
+        # Volume traded of each equity each day is a constant 100.
+        prices = list(range(1, 8))
+        volume = [100]
+
+        # Data for source panel.
+        equity0_dates = trading_days[sdi:sdi + 3]
+        equity0_data = pd.DataFrame(
+            data=zip(prices[:3], volume * 3),
+            index=equity0_dates,
+            columns=['price', 'volume'],
+        )
+
+        equity1_dates = trading_days[sdi:sdi + 5]
+        equity1_data = pd.DataFrame(
+            data=zip(prices[:5], volume * 5),
+            index=equity1_dates,
+            columns=['price', 'volume'],
+        )
+
+        equity2_dates = trading_days[sdi:sdi + 7]
+        equity2_data = pd.DataFrame(
+            data=zip(prices[:7], volume * 7),
+            index=equity2_dates,
+            columns=['price', 'volume'],
+        )
+
+        # Indexed on sid
+        cls.source_panel = pd.Panel(
+            {0: equity0_data, 1: equity1_data, 2: equity2_data},
+        )
+
+        # Initialize function shared between all test algos.
+        def initialize(context):
+            context.ordered = False
+            context.set_commission(PerShare(0))
+            context.set_slippage(FixedSlippage(spread=0))
+            context.num_positions = []
+            context.cash = []
+
+        cls.initialize = initialize
+
+        try:
+            cls.stack = ExitStack()
+            finder = cls.stack.enter_context(
+                tmp_asset_finder(equities=cls.metadata),
+            )
+
+            # Manually set the trading environment's asset finder.
+            cls.env = TradingEnvironment(asset_db_path=None)
+            cls.env.asset_finder = finder
+            cls.env.engine = finder.engine
+        except:
+            cls.tearDownClass()
+            raise
+
+    def setUp(self):
+        self.source = DataPanelSource(self.source_panel)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.stack.close()
+
+    def test_delisted_equities(self):
+        """
+        Make sure that after an equity gets delisted, our portfolio holds the
+        correct number of equities and correct amount of cash.
+        """
+        def handle_data(context, data):
```

Comment:
Each test's `handle_data` is slightly different; is it worth defining it every test, or should I try to make a shared `handle_data` like `initialize`?